# Retail Sales Data Processing and Business Insights Generation

## 1. Data Loading

In [10]:
import pandas as pd

file_path = '../data/USECASE - Data Engineering.xlsx'

product = pd.read_excel(file_path, sheet_name='product_details')
retail1 = pd.read_excel(file_path, sheet_name='retail_data1')
retail2 = pd.read_excel(file_path, sheet_name='retail_data2')

print("Product:", product.shape)
print("Retail1:", retail1.shape)
print("Retail2:", retail2.shape)


Product: (10, 4)
Retail1: (4243, 16)
Retail2: (4251, 16)


## 2. Data Exploration

In [11]:
retail1.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [12]:
retail2.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,4001,1112,Jasmine Carlson,106,80000.0,sofa,Furniture,offline,Delhi,12-18-2025,2,Cash,0.10,JasmineCarlson@gmail.com,8286662057,successful
1,4002,1125,Jeanne Suarez,105,45000.0,Tv,Electronics,online,Mumbai,2025-08-07 00:00:00,3,Card,0.10,JeanneSuarez@yahoo.com,6435924212,successful
2,4003,1379,Anthony Howard,107,50000.0,Dining table,FURN,offline,Delhi,09-19-2025,2,Cash,0.30,Anthony42@gmail.com,8187928333,successful
3,4004,1570,John Wilkins,107,50000.0,DINING TABLE,FURN,offline,Delhi,2025-11-20 00:00:00,2,Cash,0.40,JohnWilkins@outlook.com,6440039629,successful
4,4005,1433,Joseph Knapp,108,15000.0,Mixer grinder,home appliances,online,Bangalore,2025-09-29 00:00:00,5,Cash,0.25,JosephKnapp@gmail.com,9284221237,successful


In [13]:
product.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


## 3. Data Quality Assessment

In [14]:
print("Retail1 Missing Values")
print(retail1.isnull().sum())

print("\nRetail2 Missing Values")
print(retail2.isnull().sum())

Retail1 Missing Values
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                404
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

Retail2 Missing Values
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                405
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64


In [15]:
print("Retail1 Duplicates:", retail1.duplicated().sum())
print("Retail2 Duplicates:", retail2.duplicated().sum())

Retail1 Duplicates: 0
Retail2 Duplicates: 0


In [16]:
retail1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4243 entries, 0 to 4242
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4243 non-null   int64  
 1   customer_id        4243 non-null   int64  
 2   customer_name      4243 non-null   object 
 3   product_id         4243 non-null   int64  
 4   price              3839 non-null   float64
 5   product_name       4243 non-null   object 
 6   category           4243 non-null   object 
 7   purchase_location  4243 non-null   object 
 8   city               4243 non-null   object 
 9   transaction_date   4243 non-null   object 
 10  quantity           4243 non-null   int64  
 11  payment_method     4243 non-null   object 
 12  discount           4243 non-null   float64
 13  email              4243 non-null   object 
 14  phone              4243 non-null   int64  
 15  payment_status     4243 non-null   object 
dtypes: float64(2), int64(5),

In [17]:
retail2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4251 entries, 0 to 4250
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4251 non-null   int64  
 1   customer_id        4251 non-null   int64  
 2   customer_name      4251 non-null   object 
 3   product_id         4251 non-null   int64  
 4   price              3846 non-null   float64
 5   product_name       4251 non-null   object 
 6   category           4251 non-null   object 
 7   purchase_location  4251 non-null   object 
 8   city               4251 non-null   object 
 9   transaction_date   4251 non-null   object 
 10  quantity           4251 non-null   int64  
 11  payment_method     4251 non-null   object 
 12  discount           4251 non-null   float64
 13  email              4251 non-null   object 
 14  phone              4251 non-null   int64  
 15  payment_status     4251 non-null   object 
dtypes: float64(2), int64(5),

### Data Quality Issues Identified

1. Missing Values
   - Retail Data 1 contains 404 missing values in the price column.
   - Retail Data 2 contains 405 missing values in the price column.

2. Inconsistent Category Names
   - Examples: Electronics/ELEC, Furniture/FURN, Home Appliances/home appliances.

3. Inconsistent Product Names
   - Examples: TV/Tv, Sofa/sofa, Mixer Grinder/Mixer grinder.

4. Inconsistent Date Formats
   - Multiple date formats are present across datasets.

5. Personally Identifiable Information (PII)
   - Customer email and phone number fields contain sensitive information.

6. Data Type Issues
   - transaction_date is stored as object instead of datetime.

7. Duplicate Records
   - No complete duplicate records found in either dataset.

## 4. Data Integration

In [18]:
retail_raw = pd.concat([retail1, retail2], ignore_index=True)

print("Combined Dataset Shape:", retail_raw.shape)

Combined Dataset Shape: (8494, 16)


In [19]:
retail_raw.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [20]:
retail_raw.tail()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
8489,7996,1377,Steven Keith,104,8000.0,SHOES,Clothing,online,Delhi,2025-05-22 00:00:00,1,Cash,0.30,Steven25@gmail.com,7490253101,successful
8490,7997,1338,Brian Guzman,107,50000.0,Dining table,furniture,online,Chennai,2025-07-07 00:00:00,4,UPI,0.05,Brian56@gmail.com,9118813893,successful
8491,7998,1495,Ashley Mitchell,107,50000.0,Dining table,FURN,offline,Mumbai,2026-12-03 00:00:00,3,Card,0.15,AshleyMitchell@gmail.com,7994304134,successful
8492,7999,1363,Seth Everett,102,70000.0,phone,Electronics,offline,Mumbai,2025-07-21 00:00:00,3,NetBanking,0.30,SethEverett@gmail.com,8307793839,successful
8493,8000,1650,Andre Lamb,110,45000.0,Microwave,HOME,online,Delhi,2025-02-06 00:00:00,4,UPI,0.20,Andre49@gmail.com,6958582191,successful


## 5. Data Cleaning & Transformation

In [21]:
retail_clean = retail_raw.copy()

print(retail_clean.shape)

(8494, 16)


In [22]:
sorted(retail_clean['category'].unique())

['CLOTH',
 'Clothing',
 'ELEC',
 'Electronics',
 'FURN',
 'Furniture',
 'HOME',
 'Home Appliances',
 'clothing',
 'electronics',
 'furniture',
 'home appliances']

In [23]:
category_mapping = {
    'ELEC': 'Electronics',
    'electronics': 'Electronics',

    'FURN': 'Furniture',
    'furniture': 'Furniture',

    'CLOTH': 'Clothing',
    'clothing': 'Clothing',

    'HOME': 'Home Appliances',
    'home appliances': 'Home Appliances'
}

retail_clean['category'] = retail_clean['category'].replace(category_mapping)

In [24]:
sorted(retail_clean['category'].unique())

['Clothing', 'Electronics', 'Furniture', 'Home Appliances']

In [25]:
sorted(retail_clean['product_name'].unique())

['DINING TABLE',
 'Dining table',
 'LAPTOP',
 'Laptop',
 'MICROWAVE',
 'MIXER GRINDER',
 'Microwave',
 'Mixer grinder',
 'PHONE',
 'Phone',
 'REFRIGERATOR',
 'Refrigerator',
 'SHIRT',
 'SHOES',
 'SOFA',
 'Shirt',
 'Shoes',
 'Sofa',
 'TV',
 'Tv',
 'dining table',
 'laptop',
 'microwave',
 'mixer grinder',
 'phone',
 'refrigerator',
 'shirt',
 'shoes',
 'sofa',
 'tv']

In [26]:
retail_clean['product_name'] = retail_clean['product_name'].str.title()

In [27]:
sorted(retail_clean['product_name'].unique())

['Dining Table',
 'Laptop',
 'Microwave',
 'Mixer Grinder',
 'Phone',
 'Refrigerator',
 'Shirt',
 'Shoes',
 'Sofa',
 'Tv']

In [28]:
sorted(product['product_name'].unique())

['Dining Table',
 'Laptop',
 'Microwave',
 'Mixer Grinder',
 'Phone',
 'Refrigerator',
 'Shirt',
 'Shoes',
 'Sofa',
 'TV']

In [29]:
retail_clean['price'].isnull().sum()

809

In [30]:
retail_clean[retail_clean['price'].isnull()].head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
9,9,749,Linda Lara,110,NaN,Microwave,Home Appliances,offline,Mumbai,2025-08-30 00:00:00,3,Card,0.30,LindaLara@outlook.com,7476164201,successful
17,17,570,Donald Carter,102,NaN,Phone,Electronics,offline,Chennai,2025-05-19 00:00:00,3,NetBanking,0.15,Donald21@gmail.com,7944230219,successful
19,19,786,Jeremiah Simmons,103,NaN,Shirt,Clothing,offline,Chennai,2025-09-07 00:00:00,4,Cash,0.20,JeremiahSimmons@outlook.com,7139465116,successful
23,23,104,Amanda Castillo,108,NaN,Mixer Grinder,Home Appliances,online,Hyderabad,2026-04-15 00:00:00,2,UPI,0.25,AmandaCastillo@yahoo.com,9259539548,successful
61,60,874,Samuel Patel,106,NaN,Sofa,Furniture,offline,Mumbai,05-20-2025,5,UPI,0.25,Samuel51@gmail.com,6800413067,successful


In [31]:
retail_clean[retail_clean['price'].isnull()][['product_id','product_name','price']].head(10)

,product_id,product_name,price
9,110,Microwave,NaN
17,102,Phone,NaN
19,103,Shirt,NaN
23,108,Mixer Grinder,NaN
61,106,Sofa,NaN
78,110,Microwave,NaN
85,110,Microwave,NaN
95,108,Mixer Grinder,NaN
99,110,Microwave,NaN
109,105,Tv,NaN


In [32]:
price_mapping = product.set_index('product_id')['price']

price_mapping

product_id
101    250000
102     70000
103      1300
104      8000
105     45000
106     80000
107     50000
108     15000
109     65000
110     45000
Name: price, dtype: int64

In [33]:
retail_clean['price'] = retail_clean['price'].fillna(
    retail_clean['product_id'].map(price_mapping)
)

In [34]:
retail_clean['price'].isnull().sum()

0

Missing prices were not imputed using statistical measures such as mean or median. Instead, missing values were populated using the standardized product price available in the product reference table (product_details), ensuring business consistency and data accuracy.

### Date Standardization

In [35]:
retail_clean['transaction_date'].dtype

dtype('O')

In [36]:
retail_clean['transaction_date'] = pd.to_datetime(
    retail_clean['transaction_date'],
    format='mixed'
)

In [37]:
retail_clean['transaction_date'].dtype

dtype('<M8[ns]')

### Quantity Validation

In [38]:
retail_clean[retail_clean['quantity'] <= 0]

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
33,33,50,Stacey Henderson,104,8000.0,Shoes,Clothing,online,Mumbai,2025-06-14,-1,Card,0.40,Stacey33@gmail.com,6033182385,successful
75,74,907,Jeffrey Mccall,102,70000.0,Phone,Electronics,online,Hyderabad,2025-05-26,0,Card,0.40,Jeffrey63@gmail.com,6709574650,successful
77,76,335,Jacob Evans,103,1300.0,Shirt,Clothing,offline,Mumbai,2026-04-11,-2,NetBanking,0.35,JacobEvans@yahoo.com,6483959310,successful
78,77,780,Elizabeth Miller,110,45000.0,Microwave,Home Appliances,offline,Delhi,2025-09-14,0,Card,0.30,ElizabethMiller@yahoo.com,8279569798,successful
100,97,723,Christopher Walter,106,80000.0,Sofa,Furniture,offline,Hyderabad,2026-02-27,0,NetBanking,0.15,ChristopherWalter@gmail.com,9036645707,successful
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8113,7640,1138,Darren Hall,107,50000.0,Dining Table,Furniture,offline,Hyderabad,2025-03-09,-1,Card,0.40,DarrenHall@outlook.com,7899737394,successful
8393,7908,1759,Tamara Carter,102,70000.0,Phone,Electronics,online,Hyderabad,2025-10-30,-2,Cash,0.35,TamaraCarter@outlook.com,6536528655,successful
8432,7944,1353,Tonya Johnson,102,70000.0,Phone,Electronics,online,Delhi,2025-09-15,-2,Card,0.10,Tonya48@gmail.com,9931842216,successful
8440,7951,1554,Darrell Watson,103,1300.0,Shirt,Clothing,online,Mumbai,2025-10-27,-2,Cash,0.05,DarrellWatson@gmail.com,9492817032,successful


In [39]:
(retail_clean['quantity'] <= 0).sum()

93

In [40]:
retail_clean[retail_clean['price'] <= 0]

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status


In [41]:
(retail_clean['price'] <= 0).sum()

0

### Removing Invalid Quantity Records

In [42]:
before_rows = retail_clean.shape[0]

retail_clean = retail_clean[retail_clean['quantity'] > 0]

after_rows = retail_clean.shape[0]

print("Rows Before:", before_rows)
print("Rows After :", after_rows)
print("Rows Removed:", before_rows - after_rows)

Rows Before: 8494
Rows After : 8401
Rows Removed: 93


In [43]:
(retail_clean['quantity'] <= 0).sum()

0

In [44]:
retail_clean.to_csv(
    '../output/retail_clean_checkpoint.csv',
    index=False
)

## 6. PII Masking

In [45]:
retail_clean[['email','phone']].head()

,email,phone
0,Troy60@gmail.com,8385276968
1,SarahGuerrero@yahoo.com,7147248911
2,SamanthaHull@outlook.com,7415321565
3,GeraldCooper@gmail.com,9739354201
4,CameronBlack@yahoo.com,9938169477


In [46]:
def mask_email(email):
    username, domain = email.split('@')
    return username[0] + '*' * (len(username)-1) + '@' + domain

retail_clean['email_masked'] = retail_clean['email'].apply(mask_email)

In [47]:
retail_clean[['email','email_masked']].head()

,email,email_masked
0,Troy60@gmail.com,T*****@gmail.com
1,SarahGuerrero@yahoo.com,S************@yahoo.com
2,SamanthaHull@outlook.com,S***********@outlook.com
3,GeraldCooper@gmail.com,G***********@gmail.com
4,CameronBlack@yahoo.com,C***********@yahoo.com


In [48]:
def mask_phone(phone):
    phone = str(phone)
    return phone[:2] + '*' * 6 + phone[-2:]

retail_clean['phone_masked'] = retail_clean['phone'].apply(mask_phone)

In [49]:
retail_clean[['phone','phone_masked']].head()

,phone,phone_masked
0,8385276968,83******68
1,7147248911,71******11
2,7415321565,74******65
3,9739354201,97******01
4,9938169477,99******77


In [50]:
retail_clean.drop(
    columns=['email','phone'],
    inplace=True
)

In [51]:
retail_clean.columns

Index(['transaction_id', 'customer_id', 'customer_name', 'product_id', 'price',
       'product_name', 'category', 'purchase_location', 'city',
       'transaction_date', 'quantity', 'payment_method', 'discount',
       'payment_status', 'email_masked', 'phone_masked'],
      dtype='object')

## 7. Product Dimension Enrichment

In [52]:
retail_enriched = retail_clean.merge(
    product,
    on='product_id',
    how='left',
    suffixes=('_txn', '_master')
)

retail_enriched.shape

(8401, 19)

In [53]:
retail_enriched.head()

,transaction_id,customer_id,customer_name,product_id,price_txn,product_name_txn,category_txn,purchase_location,city,transaction_date,quantity,payment_method,discount,payment_status,email_masked,phone_masked,product_name_master,category_master,price_master
0,1,642,Troy Mitchell,108,15000.0,Mixer Grinder,Home Appliances,offline,Bangalore,2025-12-06,3,UPI,0.10,successful,T*****@gmail.com,83******68,Mixer Grinder,Home Appliances,15000
1,2,881,Sarah Guerrero,102,70000.0,Phone,Electronics,online,Bangalore,2025-07-21,5,Card,0.35,successful,S************@yahoo.com,71******11,Phone,Electronics,70000
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11,2,UPI,0.05,successful,S***********@outlook.com,74******65,Refrigerator,Home Appliances,65000
3,4,794,Gerald Cooper,106,80000.0,Sofa,Furniture,offline,Chennai,2025-10-06,2,UPI,0.05,successful,G***********@gmail.com,97******01,Sofa,Furniture,80000
4,5,864,Cameron Black,108,15000.0,Mixer Grinder,Home Appliances,offline,Chennai,2025-12-12,1,Cash,0.05,successful,C***********@yahoo.com,99******77,Mixer Grinder,Home Appliances,15000


## 8. KPI Generation

In [54]:
retail_enriched['revenue'] = (
    retail_enriched['price_txn']
    * retail_enriched['quantity']
    * (1 - retail_enriched['discount'])
)

In [55]:
retail_enriched[['price_txn','quantity','discount','revenue']].head()

,price_txn,quantity,discount,revenue
0,15000.0,3,0.10,40500.0
1,70000.0,5,0.35,227500.0
2,65000.0,2,0.05,123500.0
3,80000.0,2,0.05,152000.0
4,15000.0,1,0.05,14250.0


KPI 1: Total Revenue

In [56]:
total_revenue = retail_enriched['revenue'].sum()

print(f"Total Revenue: ₹{total_revenue:,.2f}")

Total Revenue: ₹1,239,132,235.00


KPI 2: Revenue by Category

In [57]:
revenue_by_category = (
    retail_enriched
    .groupby('category_txn')['revenue']
    .sum()
    .sort_values(ascending=False)
)

revenue_by_category

category_txn
Electronics        721682500.0
Furniture          257064000.0
Home Appliances    242328000.0
Clothing            18057735.0
Name: revenue, dtype: float64

KPI 3: Revenue by City

In [58]:
revenue_by_city = (
    retail_enriched
    .groupby('city')['revenue']
    .sum()
    .sort_values(ascending=False)
)

revenue_by_city

city
Chennai      259249060.0
Delhi        256872880.0
Hyderabad    248807740.0
Mumbai       247071395.0
Bangalore    227131160.0
Name: revenue, dtype: float64

KPI 4: Top Selling Products

In [59]:
top_products = (
    retail_enriched
    .groupby('product_name_txn')['revenue']
    .sum()
    .sort_values(ascending=False)
)

top_products.head(10)

product_name_txn
Laptop           500675000.0
Sofa             158244000.0
Phone            134760500.0
Refrigerator     124390500.0
Dining Table      98820000.0
Microwave         89214750.0
Tv                86247000.0
Mixer Grinder     28722750.0
Shoes             15632000.0
Shirt              2425735.0
Name: revenue, dtype: float64

KPI 5: Revenue by Payment Method

In [60]:
revenue_by_payment = (
    retail_enriched
    .groupby('payment_method')['revenue']
    .sum()
    .sort_values(ascending=False)
)

revenue_by_payment

payment_method
Card          324934475.0
Cash          318279145.0
UPI           307074335.0
NetBanking    288844280.0
Name: revenue, dtype: float64

KPI 6: Monthly Revenue Trend

In [61]:
monthly_revenue = (
    retail_enriched
    .groupby(retail_enriched['transaction_date'].dt.to_period('M'))['revenue']
    .sum()
)

monthly_revenue

transaction_date
2025-01      9720350.0
2025-02      7876940.0
2025-03      8102655.0
2025-04     61771055.0
2025-05     99549940.0
2025-06     98937645.0
2025-07     98422095.0
2025-08    105666830.0
2025-09    102679395.0
2025-10    102449430.0
2025-11     95349495.0
2025-12     92427350.0
2026-01     93221550.0
2026-02     98139805.0
2026-03     87566585.0
2026-04     39921435.0
2026-05      2738300.0
2026-06      6428535.0
2026-07      7891595.0
2026-08      4360270.0
2026-09      2621700.0
2026-10      4843855.0
2026-11      3707050.0
2026-12      4738375.0
Freq: M, Name: revenue, dtype: float64

In [62]:
retail_enriched.to_csv(
    '../output/retail_curated.csv',
    index=False
)

print("Curated dataset saved successfully!")

Curated dataset saved successfully!


In [63]:
retail_enriched.columns

Index(['transaction_id', 'customer_id', 'customer_name', 'product_id',
       'price_txn', 'product_name_txn', 'category_txn', 'purchase_location',
       'city', 'transaction_date', 'quantity', 'payment_method', 'discount',
       'payment_status', 'email_masked', 'phone_masked', 'product_name_master',
       'category_master', 'price_master', 'revenue'],
      dtype='object')

In [64]:
curated_data = retail_enriched[
    [
        'transaction_id',
        'customer_id',
        'customer_name',
        'product_id',
        'product_name_txn',
        'category_txn',
        'city',
        'purchase_location',
        'transaction_date',
        'quantity',
        'price_txn',
        'discount',
        'payment_method',
        'payment_status',
        'email_masked',
        'phone_masked',
        'revenue'
    ]
].copy()

In [65]:
curated_data.rename(columns={
    'product_name_txn': 'product_name',
    'category_txn': 'category',
    'price_txn': 'price'
}, inplace=True)

In [66]:
curated_data.head()

,transaction_id,customer_id,customer_name,product_id,product_name,category,city,purchase_location,transaction_date,quantity,price,discount,payment_method,payment_status,email_masked,phone_masked,revenue
0,1,642,Troy Mitchell,108,Mixer Grinder,Home Appliances,Bangalore,offline,2025-12-06,3,15000.0,0.10,UPI,successful,T*****@gmail.com,83******68,40500.0
1,2,881,Sarah Guerrero,102,Phone,Electronics,Bangalore,online,2025-07-21,5,70000.0,0.35,Card,successful,S************@yahoo.com,71******11,227500.0
2,3,505,Samantha Hull,109,Refrigerator,Home Appliances,Chennai,offline,2025-07-11,2,65000.0,0.05,UPI,successful,S***********@outlook.com,74******65,123500.0
3,4,794,Gerald Cooper,106,Sofa,Furniture,Chennai,offline,2025-10-06,2,80000.0,0.05,UPI,successful,G***********@gmail.com,97******01,152000.0
4,5,864,Cameron Black,108,Mixer Grinder,Home Appliances,Chennai,offline,2025-12-12,1,15000.0,0.05,Cash,successful,C***********@yahoo.com,99******77,14250.0


In [67]:
curated_data.columns

Index(['transaction_id', 'customer_id', 'customer_name', 'product_id',
       'product_name', 'category', 'city', 'purchase_location',
       'transaction_date', 'quantity', 'price', 'discount', 'payment_method',
       'payment_status', 'email_masked', 'phone_masked', 'revenue'],
      dtype='object')

In [68]:
curated_data.to_csv(
    '../output/retail_curated_final.csv',
    index=False
)

print("Final curated dataset saved!")

Final curated dataset saved!


## 9. Final Dataset Validation

In [69]:
print("Final Shape:", curated_data.shape)

print("\nMissing Values:")
print(curated_data.isnull().sum())

print("\nDuplicate Records:")
print(curated_data.duplicated().sum())

Final Shape: (8401, 17)

Missing Values:
transaction_id       0
customer_id          0
customer_name        0
product_id           0
product_name         0
category             0
city                 0
purchase_location    0
transaction_date     0
quantity             0
price                0
discount             0
payment_method       0
payment_status       0
email_masked         0
phone_masked         0
revenue              0
dtype: int64

Duplicate Records:
0


In [70]:
import os

os.listdir('../output')

['retail_clean_checkpoint.csv',
 'retail_curated.csv',
 'retail_curated_final.csv']